File: qbopt/optimize/edges.py
"""Place semantic operations on conditional edges, not on either arm."""

from dataclasses import replace

from qbopt.model import ir, mir


def explicit(block: mir.MirBlock, target: int) -> bool:
    return conditional(block, target) and block.ops[-1].target == target


def conditional(block: mir.MirBlock, target: int) -> bool:
    return (len(block.succ) == 2 and target in block.succ and bool(block.ops)
            and block.ops[-1].kind is mir.Kind.BRANCH and block.ops[-1].target in block.succ)


def fresh(body: mir.MirBody) -> int:
    return max((body.entry + 1) << 32, max(block.at for block in body.blocks)) + 1


def split(body: mir.MirBody, source: int, target: int, label: int,
          ops: tuple[mir.Op, ...]) -> mir.MirBody:
    parent = body.block(source)
    if parent is None or not conditional(parent, target) or body.block(label) is not None:
        raise ValueError("edge split does not identify a fresh conditional edge")
    jump = mir.Op(label, ir.Operation.JUMP, "", (), (), kind=mir.Kind.JUMP,
                  target=target, symbol=False)
    bridge = mir.MirBlock(label, (), (*ops, jump), (target,))
    blocks = []
    for block in body.blocks:
        if block.at == source:
            block = replace(block, succ=tuple(label if at == target else at for at in block.succ),
                            ops=(*block.ops[:-1], replace(block.ops[-1], target=label)
                                 if block.ops[-1].target == target else block.ops[-1]))
        if block.at == target:
            block = replace(block, phis=tuple(replace(phi, incoming={
                label if at == source else at: value for at, value in phi.incoming.items()
            }) for phi in block.phis))
        blocks.append(block)
    return replace(body, blocks=(*blocks, bridge))

File: src/optimize/edges.rs
//! Port of `qbopt/optimize/edges.py`: place semantic operations on
//! conditional edges, not on either arm.

use crate::model::ir::Operation;
use crate::model::mir::{Kind, MirBlock, MirBody, Op, OpCode};

pub fn explicit(block: &MirBlock, target: i64) -> bool {
    conditional(block, target) && block.ops.last().and_then(|op| op.target) == Some(target)
}

pub fn conditional(block: &MirBlock, target: i64) -> bool {
    block.succ.len() == 2
        && block.succ.contains(&target)
        && block.ops.last().is_some_and(|last| {
            last.kind == Kind::Branch && last.target.is_some_and(|at| block.succ.contains(&at))
        })
}

pub fn fresh(body: &MirBody) -> i64 {
    ((body.entry + 1) << 32).max(body.blocks.iter().map(|block| block.at).max().expect("max() arg is an empty sequence"))
        + 1
}

pub fn split(body: &MirBody, source: i64, target: i64, label: i64, ops: Vec<Op>) -> Result<MirBody, String> {
    let parent = body.block(source);
    if parent.is_none_or(|parent| !conditional(parent, target)) || body.block(label).is_some() {
        return Err("edge split does not identify a fresh conditional edge".into());
    }
    let mut jump = Op::new(label, OpCode::Operation(Operation::Jump), "", vec![], vec![]);
    jump.kind = Kind::Jump;
    jump.target = Some(target);
    jump.symbol = Some(false);
    let mut bridge_ops = ops;
    bridge_ops.push(jump);
    let bridge = MirBlock::new(label, vec![], bridge_ops, vec![target]);
    let mut blocks = Vec::new();
    for block in &body.blocks {
        let mut block = block.clone();
        if block.at == source {
            block.succ = block.succ.iter().map(|&at| if at == target { label } else { at }).collect();
            let last = block.ops.last_mut().expect("a conditional edge ends in a branch");
            if last.target == Some(target) {
                last.target = Some(label);
            }
        }
        if block.at == target {
            for phi in &mut block.phis {
                phi.incoming = phi
                    .incoming
                    .iter()
                    .map(|(&at, &value)| (if at == source { label } else { at }, value))
                    .collect();
            }
        }
        blocks.push(block);
    }
    blocks.push(bridge);
    Ok(MirBody { blocks, ..body.clone() })
}
